### ISSUE TITLE
Create wiki entry for commit message style and issue/development flow

### ISSUE THREAD

**Issue by theeldestelder** [teslaworksumn/lightshow-Gallium#7]
The wiki should explain how this project expects commit messages to be styled as well as what the workflow of the project should look like as a reference to project collaborators.

**Comment by nickschatz** [comment]
I think it's silly to require commit messages or titles, and PR titles and etc, being in a specific format other than "be descriptive."

**Comment by nickschatz** [comment]
Branch names being issue names is a good idea, but any more than that is redundant.

**Comment by theeldestelder** [comment]
Requiring commit titles may seem silly, but we are requiring it for a number of reasons.

1. When we are searching through the repository in the future and are wondering how a certain issue got fixed, we can simply search for the issue number and the commit(s) associated with fixing that issue will pop up and let us easily see what changed.
2. Companies require a standard format for commit titles/messages and we are hoping to emulate at least some of those practices.  Classes don't really teach collaboration with Git and we are hoping to give people a small experience of what it's like to work with a (hopefully) maintainable codebase that follows a number of guidelines that extend to commit titles/messages.

We aren't requiring commit messages to be in a certain format other than what you mentioned earlier: be descriptive.  As long as the title contains the issue number and issue title, you can format the body however you like as long as the imperative tense is used.

[Here is a great article](https://chris.beams.io/posts/git-commit/) about why the things I mentioned are important.  Our commit titles will instead refer the specific issue number it addresses but everything else is pretty applicable.

It may seem like we're being pretty picky and you wouldn't be entirely wrong.  That said, we are hoping that requiring high quality commit messages will help the project become more scalable and maintainable in the long run.

**Comment by nickschatz** [comment]
I'll drop https://github.com/nodejs/node and https://github.com/Microsoft/vscode here. Both of these are gigantic projects that take contributions from many more contributors than this repository, and neither has any restrictions on what commit or PR titles should be.

If you want to search for PRs relating to a certain issue, all that needs to be done is reference the issue number in the PR description. This also has the benefit of automatically linking and closing the issue once the PR is merged (if it's referenced like "Closes #xx".)

 > Companies require a standard format for commit titles/messages and we are hoping to emulate at least some of those practices. Classes don't really teach collaboration with Git and we are hoping to give people a small experience of what it's like to work with a (hopefully) maintainable codebase that follows a number of guidelines that extend to commit titles/messages.

So what? Requiring commit messages be the issue title isn't teaching git. As I argue below, this tends to make the codebase *less* maintainable. If someone knows git, they can conform to these restrictions without issue, but I seriously doubt this is as widespread as you say.

I mentioned this at the meeting, but what if I need multiple commits to fix an issue? Then the title is all the same, and it's impossible to differentiate commits without expanding them. I'd say that restricting the title, even neglecting the first point, makes the git history less readable. In fact the blog post you linked is evidence for this, saying that the subject should be descriptive, and may not always require a body.

If you want suggestions for more useful contributing guidelines, since we're all working in the same repository and not forks, I'd put down these:

1. Branch titles have to be "issXX-name-of-issue". (Currently a rule)
2. Don't be shy about committing, and be descriptive in commit subjects and messages. (If you want to link or summarize the linked blogpost, that has a lot of good information)
3. Pull request titles should be descriptive, and their descriptions should contain an overview of major changes. They should also reference their issue like "Closes #xx". (Good to get a summary of what's being changed at a glance)

If you're open to everyone working in forks to their personal Github accounts, then the first rule can be omitted because we don't have to share a branch namespace. If you're looking for a "real" experience, this is how every single OSS that lives on Github does it.

And as I mentioned on my "intro" issue you can squash commits as you merge PRs. This turns even messy commit subjects into readable PR titles.

I realize I'm being very forward with this as I'm just joining the project, but I've been around the block once or twice. Ignore this if you want, I'll drop it because I've already spent way more time arguing about it than is worth, and I'll contribute under your rules.

**Comment by theeldestelder** [comment]
Thanks for the response and you've brought up a lot of good points.  Byron and I will have to talk about next steps but I anticipate some policy changes in the light of this thread because right now it is pretty imposing and complicated to contribute.  I've never done this before and appreciate your patient responses.  Hopefully we'll get back to you by the end of the day with some updated policies and we'll make those clear both to you and everyone else hoping to contribute.  Thanks again!

**Comment by nickschatz** [comment]
Thank you for being open about this. I agree that having the issue in commit titles makes it easy to reference. I look forward to working with you all on this project.

Onto the issue of code style, on a different (Java) project I worked with tools called PMD and Checkstyle, which automated some of the style and code quality checking. PMD works with Javascript, but seems to share a lot of checks with JSLint. I can't seem to find a Javascript equivalent to Checkstyle, which did stuff like enforce periods at the end of comments and limit number of lines in each function, which did wonders for readability. Of course this can also be done in code reviews but it's a bit less tedious to have a tool do it.

**Comment by ByronAmbright** [comment]
Currently we are using `eslint` to lint the code with AirBnB javascript coding style. It should be run on a pipeline every time a new pull request is opened or a commit is pushed to a pull request.  I am open to changing this. I believe `eslint` will catch most things like unused variables, semi colons or spacing in the code. I do not think it looks at comments however. PMD might catch somethings that eslint does not catch. We should test this.
